File: loffice-viewer/app/src/main/java/com/lofficeviewer/app/data/DocumentFormat.kt
package com.lofficeviewer.app.data

enum class DocumentFormat(
    val label: String,
    val extensions: List<String>,
) {
    PDF("PDF", listOf(".pdf")),
    TXT("텍스트", listOf(".txt", ".md", ".markdown", ".csv", ".log", ".json", ".xml")),
    DOCX("Word", listOf(".docx", ".doc")),
    PPTX("PowerPoint", listOf(".pptx", ".ppt")),
    XLSX("Excel", listOf(".xlsx", ".xls")),
    HWP("HWP", listOf(".hwp")),
    HWPX("HWPX", listOf(".hwpx")),
    UNKNOWN("기타", emptyList());

    companion object {
        fun fromFileName(fileName: String): DocumentFormat {
            val lower = fileName.lowercase()
            return entries.firstOrNull { fmt ->
                fmt.extensions.any { lower.endsWith(it) }
            } ?: UNKNOWN
        }

        fun fromMimeType(mime: String?): DocumentFormat? {
            if (mime.isNullOrBlank()) return null
            return when {
                mime.contains("pdf") -> PDF
                mime.contains("plain") -> TXT
                mime.contains("wordprocessingml") || mime == "application/msword" -> DOCX
                mime.contains("presentationml") || mime.contains("powerpoint") -> PPTX
                mime.contains("spreadsheetml") || mime.contains("excel") -> XLSX
                mime.contains("hwp") || mime.contains("hancom") -> if (mime.contains("hwpx") || mime.contains("hwp+zip")) HWPX else HWP
                else -> null
            }
        }
    }
}

data class ViewerDocument(
    val uriString: String,
    val displayName: String,
    val format: DocumentFormat,
)

File: loffice-viewer-ios/LofficeViewer/Models/DocumentFormat.swift
import Foundation

enum DocumentFormat: String, Codable, CaseIterable {
    case pdf = "PDF"
    case txt = "텍스트"
    case docx = "Word"
    case pptx = "PowerPoint"
    case xlsx = "Excel"
    case hwp = "HWP"
    case hwpx = "HWPX"
    case unknown = "기타"

    var label: String { rawValue }

    static func from(fileName: String) -> DocumentFormat {
        let lower = fileName.lowercased()
        if lower.hasSuffix(".pdf") { return .pdf }
        if [".txt", ".md", ".markdown", ".csv", ".log", ".json", ".xml"].contains(where: { lower.hasSuffix($0) }) { return .txt }
        if lower.hasSuffix(".docx") || lower.hasSuffix(".doc") { return .docx }
        if lower.hasSuffix(".pptx") || lower.hasSuffix(".ppt") { return .pptx }
        if lower.hasSuffix(".xlsx") || lower.hasSuffix(".xls") { return .xlsx }
        if lower.hasSuffix(".hwpx") { return .hwpx }
        if lower.hasSuffix(".hwp") { return .hwp }
        return .unknown
    }
}

struct ViewerDocument: Identifiable, Codable, Equatable {
    var id: String { url.absoluteString }
    let url: URL
    let displayName: String
    let format: DocumentFormat

    enum CodingKeys: String, CodingKey {
        case url, displayName, format
    }
}

@MainActor
final class DocumentStore: ObservableObject {
    @Published private(set) var recent: [ViewerDocument] = []
    private let key = "loffice.recent"

    init() {
        load()
    }

    func addRecent(_ doc: ViewerDocument) {
        recent.removeAll { $0.url == doc.url }
        recent.insert(doc, at: 0)
        recent = Array(recent.prefix(20))
        save()
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: key),
              let decoded = try? JSONDecoder().decode([ViewerDocument].self, from: data) else { return }
        recent = decoded
    }

    private func save() {
        if let data = try? JSONEncoder().encode(recent) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }
}
